packages/contracts-bedrock/src/cannon/MIPS64.sol

```diff
@@ -0,0 +1,930 @@
+            if (opcode == 0 && fun == 0xC) {
+                return handleSyscall(_localContext);
+            }
+
+            // Handle RMW (read-modify-write) ops
+            if (opcode == ins.OP_LOAD_LINKED || opcode == ins.OP_STORE_CONDITIONAL) {
+                return handleRMWOps(state, thread, insn, opcode);
+            }
+            if (opcode == ins.OP_LOAD_LINKED64 || opcode == ins.OP_STORE_CONDITIONAL64) {
+                return handleRMWOps(state, thread, insn, opcode);
+            }
+
+            // Exec the rest of the step logic
+            st.CpuScalars memory cpu = getCpuScalars(thread);
+            ins.CoreStepLogicParams memory coreStepArgs = ins.CoreStepLogicParams({
+                cpu: cpu,
+                registers: thread.registers,
+                memRoot: state.memRoot,
+                memProofOffset: MIPS64Memory.memoryProofOffset(MEM_PROOF_OFFSET, 1),
+                insn: insn,
+                opcode: opcode,
+                fun: fun
+            });
+            bool memUpdated;
+            uint64 memAddr;
+            (state.memRoot, memUpdated, memAddr) = ins.execMipsCoreStepLogic(coreStepArgs);
+            setStateCpuScalars(thread, cpu);
+            updateCurrentThreadRoot();
+            if (memUpdated) {
+                handleMemoryUpdate(state, memAddr);
+            }
+
+            return outputState();
+        }
+    }
+
+    function handleMemoryUpdate(State memory _state, uint64 _memAddr) internal pure {
+        if (_memAddr == (ins.ADDRESS_MASK & _state.llAddress)) {
+            // Reserved address was modified, clear the reservation
+            clearLLMemoryReservation(_state);
+        }
+    }
+
+    function clearLLMemoryReservation(State memory _state) internal pure {
+        _state.llReservationStatus = LL_STATUS_NONE;
+        _state.llAddress = 0;
+        _state.llOwnerThread = 0;
+    }
+
+    function handleRMWOps(
+        State memory _state,
+        ThreadState memory _thread,
+        uint32 _insn,
+        uint32 _opcode
+    )
+        internal
+        returns (bytes32)
+    {
+        unchecked {
+            uint64 base = _thread.registers[(_insn >> 21) & 0x1F];
+            uint32 rtReg = (_insn >> 16) & 0x1F;
+            uint64 addr = base + ins.signExtendImmediate(_insn);
+
+            // Determine some opcode-specific parameters
+            uint8 targetStatus = LL_STATUS_ACTIVE_32_BIT;
+            uint64 byteLength = 4;
+            if (_opcode == ins.OP_LOAD_LINKED64 || _opcode == ins.OP_STORE_CONDITIONAL64) {
+                // Use 64-bit params
+                targetStatus = LL_STATUS_ACTIVE_64_BIT;
+                byteLength = 8;
+            }
+
+            uint64 retVal = 0;
+            uint64 threadId = _thread.threadID;
+            if (_opcode == ins.OP_LOAD_LINKED || _opcode == ins.OP_LOAD_LINKED64) {
+                retVal = loadWord(_state, addr);
+                retVal = ins.selectSubWord(addr, retVal, byteLength, true);
```

Comment:
Will do this but implemented in MIPS64.sol since it requires access so the function can also handle the full word load. Otherwise, the MEM_PROOF_OFFSET will have to be passed to `MIPS64Memory.sol` which inflates the stack a bit.